[core/model/base] Reset the model without recreating the objects

Dav1dK opened the pull request:
Thus the observers of the model don't need to resubscribe to all tanks, pipes and motors.
Also fix a bug, where the motor hasn't notified the observers.

Dav1dK committed:
[core/model/base] Add test for ProductionSite.{get,set}InputTemperature

Also fix a bug where the enum map wasn't initialized.

Dav1dK committed:
[core/model/base] Add Motor.reset()

Dav1dK committed:
[core/model/base] Add the default color to TankSelector

Dav1dK committed:
[core/model/base] Add Pipe.reset() and a constructor

The constructor can be used to set a default threshold different than
100, which is used for reset.

Dav1dK committed:
[core/model/base] Add a reset() method to the tanks

Dav1dK committed:
[core/model/base] Reset the ProductionSide without recreating the tanks

Otherwise all subscibed objects need to resubscribe to the tanks.
Instead just reset all values. Therefore pass a sensible default value
to Pipe.

Dav1dK committed:
[core/model/base] Fix checkstyle issue

Dav1dK committed:
[core/model/base] Notify observers if motor speed changes

Also add a testcase to verify observing the motor works.

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/AbstractTank.java:
```diff
@@ -27,9 +28,10 @@ public AbstractTank(float capacity, TankSelector tankSelector, Liquid liquid, Pi
         this.capacity = capacity;
         this.selector = tankSelector;
-        this.liquid = liquid;
+        this.liquid = liquid.clone();
```
War nicht die Idee hinter einem Datentyp ohne Setter, dass man ihn eben nicht klonen muss? Sollte eigentlich auch ohne gehen, ich meine das Liquid wird ja nicht bearbeitet

Dav1dK replied:
Dachte ich auch. Blöderweise kann die Color von Liquid bearbeitet werden. Aber vielleicht kann man das auch ohne Probleme rausnehmen, ich werds mal ausprobieren.

ByteHamster replied:
Uh stimmt... Das ist ja doof ^^

Dav1dK replied:
Ich habe Color immutable gemacht.

ByteHamster replied:
Huh und das wirft keine Fehler überall? :D

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/Liquid.java:
```diff
@@ -95,4 +95,19 @@ public int hashCode() {
+            tmp = (Liquid) super.clone();
+        } catch (CloneNotSupportedException ex) {
+            /* Should never happen */
+            throw new RuntimeException("Error in Color.clone()");
```
Color? Du machst doch super

ByteHamster replied:
Wirfst du das in Color nicht auch?

Dav1dK replied:
Ohh, CopyPaste Fehler

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/Pipe.java:
```diff
@@ -28,7 +29,26 @@ public Pipe(float crosssection, int length) {
+     * @param defaultThreshold The default threshold
+     * @throws IllegalArgumentException if crosssection or length are negative or zero.
+     */
+    public Pipe(float crosssection, int length, byte defaultThreshold) {
```
Bitte für eine der beiden entscheiden. Sonst taucht irgendwann plötzlich eine 100 auf und keiner weiß, wo sie herkommt

`defaultThreshold = (byte) 100;`

Ich meine das default sollte auf jeden Fall nur an einer Stelle sein. Entweder es wird von außen entschieden oder von innen.

Dav1dK replied:
Das Problem ist, dass sehr viele Stellen angepasst werden müssen, wenn man den Construktor ändert. Andererseits ist es aber nötig, den defaultThreshold zu setzen damit man die Pipe resetten kann.

ByteHamster replied:
Muss das wirklich so viel sein? Pipes werden doch eh nur im Model erstellt

Dav1dK replied:
Hmm stimmt. Habe gerade mal danach gesucht, es werden nur an 23 Stellen Pipes erstellt. Der Compilerfehler sah größer aus :D Dann werde ich das noch ändern.

ByteHamster replied:
Normal mag ich solche Änderungen ja nicht, aber in dem Fall wäre es sonst zu inkonsistent

Dav1dK replied:
Ich habe den ersten Konstructor jetzt entfernt.

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/ProductionSite.java:
```diff
@@ -12,8 +12,8 @@
-    protected EnumMap<TankSelector, Tank> tanks;
-    protected EnumMap<TankSelector, Float> inputTemperature;
+    protected EnumMap<TankSelector, Tank> tanks = new EnumMap<>(TankSelector.class);;
+    protected EnumMap<TankSelector, Float> inputTemperature = new EnumMap<TankSelector, Float>(TankSelector.class);
```
Final?

Dav1dK replied:
Hab ich geändert.

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/ProductionSite.java:
```diff
@@ -80,25 +80,25 @@ private void initTanks() {
             if (!specialPipeAssigned) {
-                inPipe.setValveThreshold((byte) 34);
-                outPipe.setValveThreshold((byte) 34);
+                threshold = (byte) 34;
```
Wah, genau das. Jetzt widerspricht es sich mit dem tank-internen default Wert

Dav1dK replied:
Wie meinst du das?

ByteHamster replied:
Na ja du initialisierst mit 34 und default vom pipe konstruktor ist 100

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/ProductionSite.java:
```diff
@@ -80,25 +80,25 @@ private void initTanks() {
+                threshold = (byte) 33;
             }
+            Liquid l = new Liquid(halfFull, SimulationConstants.MIN_TEMPERATURE, selector.getInitialColor());
+            Pipe inPipe = new Pipe(SimulationConstants.PIPE_CROSSSECTION, SimulationConstants.PIPE_LENGTH, threshold);
```
Ich hab so das Gefühl, dass der Zustand hier nicht ausbalanciert ist. Sollte nicht aus dem unteren Tank 3x so viel ablaufen wie aus den oberen?

Dav1dK replied:
Das tut es auch. Die Pipe vom MixTank nach unten hat den Threshold 100, die drei oberen Tanks haben 34 + 33 + 33 = 100.

ByteHamster replied:
ah okay. Hab nur nirgends die 100 gesehen auf den ersten Blick. Oder benutzt du da dann das automatische, interne? Falls ja, muss ich mich deutlich dagegen aussprechen :D

Dav1dK replied:
Zeile 84, explizit ;)

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/ProductionSite.java:
```diff
@@ -80,25 +80,25 @@ private void initTanks() {
+
 
             tanks.put(selector, instantiateTank(SimulationConstants.TANK_SIZE, selector, l, outPipe, inPipe));
+            inputTemperature.put(selector, selector.getInitialTemperature());
```
Wenn du auf Nummer sicher gehen willst, dass es WIRKLICH richtig initialisiert ist, sollte hier nochmal reset stehen

Dav1dK replied:
Stimmt, dann kann ich mir dort ein bisschen was sparen.

ByteHamster commented:
Wenn du einen bug im Motor behebst, kannst du auch gleich die Farben umsortieren ;)

Dav1dK commented:
Wird gemacht ;)

Dav1dK committed:
[core/model/base] Make Color immutable

Dav1dK committed:
[core/model/base] Make fields in ProductionSite final where possible

Dav1dK committed:
[core/model/base] Remove first constructor of Pipe

Only have one constructor with the defaultThreshold argument. Adapt all
uses of Pipe to the new constructor.

Dav1dK committed:
[core/model/base] Use TankSelector.getDefaultTemperature

Also call reset() after initTanks() to make sure we're in the intended
state.

Dav1dK committed:
[core/model/base] Assign the correct default colors to the Tanks

closes #254

Dav1dK commented:
Ich habe Color immutable gemacht, und den ersten Konstruktor von Pipe entfernt. Die anderen Sachen habe ich auch gefixed.

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/TankSelector.java:
```diff
@@ -45,6 +55,14 @@
     public Color getInitialColor() {
         return initialColors[this.ordinal()].clone();
```
Color braucht man nicht mehr klonen

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/Color.java:
```diff
@@ -187,9 +145,9 @@ public Color clone() {
             /* Should never happen */
```
Dann brauchen wir doch keine clone mehr, oder?

Dav1dK replied:
Eigentlich nicht.

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/ProductionSite.java:
```diff
@@ -12,8 +12,8 @@
-    protected EnumMap<TankSelector, Tank> tanks;
-    protected EnumMap<TankSelector, Float> inputTemperature;
+    final protected EnumMap<TankSelector, Tank> tanks = new EnumMap<>(TankSelector.class);;
+    final protected EnumMap<TankSelector, Float> inputTemperature = new EnumMap<TankSelector, Float>(TankSelector.class);
```
Reihenfolge (ist glaub auch im Travis) ist `protected final`

ByteHamster commented on src/osip-model/src/test/java/edu/kit/pse/osip/core/model/base/TankSelectorTest.java:
```diff
@@ -39,8 +39,8 @@ public void testGetUpperTankCount() {
-        assertEquals(TankSelector.YELLOW.getInitialColor(), new Color(0, 1, 0));
-        assertEquals(TankSelector.MAGENTA.getInitialColor(), new Color(0, 0, 1));
+        assertEquals(TankSelector.YELLOW.getInitialColor(), new Color(0, 0, 1));
+        assertEquals(TankSelector.MAGENTA.getInitialColor(), new Color(0, 1, 0));
```
Na super auf falsches Verhalten testen hier ^^

ByteHamster commented on src/osip-model/src/test/java/edu/kit/pse/osip/core/model/behavior/FillAlarmTest.java:
```diff
@@ -28,7 +28,8 @@
         testLiquid = new Liquid(20f, 300f, new Color(0.5, 0.5, 0.5));
-        tank = new Tank(200f, TankSelector.valuesWithoutMix()[0], testLiquid, new Pipe(200f, 30), new Pipe(200f, 30));
+        tank = new Tank(200f, TankSelector.valuesWithoutMix()[0], testLiquid, new Pipe(200f,
+            30, (byte) 100), new Pipe(200f, 30, (byte) 100));
```
```
tank = new Tank(200f, TankSelector.valuesWithoutMix()[0], testLiquid, new Pipe(200f,
    30, (byte) 100), new Pipe(200f, 30, (byte) 100));
```
vs
```
tank = new Tank(200f, TankSelector.valuesWithoutMix()[0], testLiquid,
    new Pipe(200f, 30, (byte) 100), new Pipe(200f, 30, (byte) 100));
```

Ich finde, die inneren Klammern sollten nicht umgebrochen werden. Macht es einfacher lesbar

Dav1dK replied:
Ich habe die Pipes in attribute ausgelagert.

ByteHamster commented on src/osip-model-simulation/src/test/java/edu/kit/pse/osip/core/model/simulation/MixTankSimulationTest.java:
```diff
@@ -23,7 +23,8 @@
     public void testPutIn() {
         Liquid content = new Liquid(100, 350, new Color(0x003000));
-        MixTankSimulation tank = new MixTankSimulation(1000, content, new Pipe(10, 10));
+        MixTankSimulation tank = new MixTankSimulation(1000, content, new Pipe(10, 10,
```
Ich finde, die inneren Klammern sollten nicht umgebrochen werden. Macht es einfacher lesbar

Dav1dK replied:
Habe ich ebenfalls ausgelagert.

ByteHamster reviewed (changes requested):
Sonst sieht es gut aus

Dav1dK committed:
[core/model/base] Make Color not cloneable anymore

It's immutable, so cloning is not needed.

Dav1dK committed:
[core/model/base] Reduce code duplication in tests

Use attributes for objects which are needed all the time instead of
creating them everytime.

Dav1dK committed:
[core/model/base] Fix checkstyle error

Dav1dK committed:
[core/model/base] Improve TankSelectorTest

ByteHamster commented on src/osip-model/src/test/java/edu/kit/pse/osip/core/model/base/TankSelectorTest.java:
```diff
@@ -34,13 +35,25 @@ public void testGetUpperTankCount() {
+    }
+
+    /**
+     * Check whether getIntialColor() and getInitialTemperatue/( returns a value for every tank. Should not throw an
```
Falsche Klammern ;) `/(`

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/Tank.java:
```diff
@@ -35,4 +35,10 @@ public Tank(float capacity, TankSelector tankSelector, Liquid liquid, Pipe outPi
     }
+
+    @Override
+    public void reset() {
```
Wäre es nicht sinnvoll, wenn die synchronized wären?

Dav1dK replied:
Stimmt, sollte so sein.

Dav1dK committed:
[core/model/base] Make reset() methods syncronized and remove final

ByteHamster commented on src/osip-model/src/main/java/edu/kit/pse/osip/core/model/base/Motor.java:
```diff
@@ -15,17 +15,26 @@
+     */
+    public synchronized void reset() {
+        setRPM(0);
+    }
```
Bei allen anderen nimmst du die Parameter aus dem Konstruktor zum reset. Dann solltest du es hier auch machen, um konsistent zu sein

Dav1dK committed:
[core/model/base] Add initial motor speed parameter to motor constructor
